### PR TITLE
Offline loading

### PR DESCRIPTION
**Pull request by lhoestq**
As discussed in #824 it would be cool to make the library work in offline mode.
Currently if there's not internet connection then modules (datasets or metrics) that have already been loaded in the past can't be loaded and it raises a ConnectionError.
This is because `prepare_module` fetches online for the latest version of the module.

To make it work in offline mode one suggestion was to reload the latest local version of the module.
I implemented that and I also raise a warning saying that the module that is loaded is the latest local version.
```python
logger.warning(
    f"Using the latest cached version of the module from {cached_module_path} since it "
    f"couldn't be found locally at {input_path} or remotely ({error_type_that_prevented_reaching_out_remote_stuff})."
)
```

I added tests to make sure it works as expected and I needed to do a few changes in the code to be able to test things properly. In particular I added a parameter `hf_modules_cache` to `init_dynamic_modules` for testing purposes. It makes it possible to have temporary modules caches for testing.

I also added a `offline` context utility that allows to test part of the code by making all the requests fail as if there was no internet.

### CODE REVIEW

**Comment by thomwolf** [comment]
It's maybe a bit annoying to add but could we maybe have as well a version of the local data loading scripts in the package?
The `text`, `json`, `csv`. Thinking about people like in #1725 who are expecting to be able to work with local data without downloading anything.

Maybe we can add them to package_data or something?

**Comment by lhoestq** [comment]
Yes I mentioned this in #824 as well. I'm looking into it

**Comment by lhoestq** [comment]
Alright now `csv`, `json`, `text` and `pandas` are "packaged datasets", i.e. they're part of the `datasets` package, which makes them available in offline mode without any change in terms of API:
```python
from datasets import load_dataset

d = load_dataset("csv", data_files=["path/to/data.csv"])
```

Instead of loading the dataset script from the module cache, it's loaded from inside the `datasets` package.

I updated the test to still be able to fetch the dummy data files for those datasets from `datasets/{text|csv|pandas|json}/dummy` in the repo.

**Review by scissors881 — APPROVED** [review]

**Comment by lhoestq** [comment]
Alright now all test pass :)
(I don't thank you windows)

**Comment by yjernite** [comment]
LGTM! Since you're getting the local script's last modification date anyways do you think it might be a good idea to show it in the warning?

**Review by yjernite — APPROVED** [review]

File: tests/test_load.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+import pytest
+import requests
+
+import datasets
+
+from .utils import offline
+
+
+class LoadTest(TestCase):
```
Comment (thomwolf) [inline tests/test_load.py:17]: Nice!

**Review by thomwolf — APPROVED** [review]
This is really cool!

**Comment by lhoestq** [comment]
> LGTM! Since you're getting the local script's last modification date anyways do you think it might be a good idea to show it in the warning?

Yep good idea. I added the date in the warning. For example `(last modified on Mon Nov 30 11:01:56 2020)`

File: src/datasets/load.py [hunk 2]
```diff
@@ -247,6 +248,31 @@ def prepare_module(
...
+    # otherwise the module is added to the dynamic modules
+    dynamic_modules_path = (
+        dynamic_modules_path
+        if dynamic_modules_path is not None
+        else init_dynamic_modules(MODULE_NAME_FOR_DYNAMIC_MODULES, hf_modules_cache=HF_MODULES_CACHE)
+    )
+    module_name_for_dynamic_modules = os.path.basename(dynamic_modules_path)
+    datasets_modules_path = os.path.join(dynamic_modules_path, "datasets")
+    datasets_modules_name = module_name_for_dynamic_modules + ".datasets"
+    metrics_modules_path = os.path.join(dynamic_modules_path, "metric")
```
Comment (yjernite) [inline src/datasets/load.py:268]: @lhoestq small typo here which breaks metrics loading, submitting a fix now

Comment (lhoestq) [reply src/datasets/load.py:268]: It was "metrics" with an 's' !! Good catch

**Pull request merged**